### PR TITLE
Feature/gymboy

### PR DESCRIPTION
**Pull request by MathisFederico**
Following issue #124 this is a first version to create a Gym environement from a PyBoy object
There is still some tests to do ! But it's a working start !

### CODE REVIEW

File: pyboy/gymboy.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        if self.game_wrapper is None:
+            raise ValueError("You need to build a game_wrapper to use this function. Otherwise there is no way to build a reward function automaticaly (for now).")
+        self.last_fitness = self.game_wrapper.fitness
+
+        # Building the action_space
+        self._DO_NOTHING = -1
+        self._buttons = list(range(1, 9)) # UP DOWN RIGHT LEFT A B SELECT START
+        self._button_is_pressed = {button:False for button in self._buttons}
+        
+        self._buttons_release = list(range(9, 17)) # UP DOWN RIGHT LEFT A B SELECT START
```
Comment (Baekalfen) [inline pyboy/gymboy.py:87]: Same here

File: pyboy/gymboy.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        self._release_button = {button:r_button for button, r_button in zip(self._buttons, self._buttons_release)}
+        
+        self.actions = [self._DO_NOTHING] + self._buttons
+        if buttons_press_mode == 'all':
+            self.actions += self._buttons_release
+        elif not (buttons_press_mode == 'press' or buttons_press_mode == 'toggle'):
+            raise ValueError(f'buttons_press_mode {buttons_press_mode} is unknowed')
```
Comment (Baekalfen) [inline pyboy/gymboy.py:94]: > unknowed

"unknown" or "invalid"

File: pyboy/gymboy.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        self._buttons = list(range(1, 9)) # UP DOWN RIGHT LEFT A B SELECT START
+        self._button_is_pressed = {button:False for button in self._buttons}
+        
+        self._buttons_release = list(range(9, 17)) # UP DOWN RIGHT LEFT A B SELECT START
+        self._release_button = {button:r_button for button, r_button in zip(self._buttons, self._buttons_release)}
+        
+        self.actions = [self._DO_NOTHING] + self._buttons
+        if buttons_press_mode == 'all':
+            self.actions += self._buttons_release
+        elif not (buttons_press_mode == 'press' or buttons_press_mode == 'toggle'):
```
Comment (Baekalfen) [inline pyboy/gymboy.py:93]: Can also be written `elif buttons_press_mode not in ['press', 'toggle']:`

File: pyboy/gymboy.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            raise ValueError(f'buttons_press_mode {buttons_press_mode} is unknowed')
+        self.buttons_press_mode = buttons_press_mode
+
+        if simultaneous_actions:
+            raise NotImplementedError("Not implemented yet, raise an issue if needed")
```
Comment (Baekalfen) [inline pyboy/gymboy.py:98]: > raise an issue

"raise an issue on GitHub" for clarity?

File: pyboy/gymboy.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        else:
+            self.action_space = Discrete(len(self.actions))
+
+        # Building the observation_space
+        if observation_type == 'raw':
+            screen = np.asarray(self.game.botsupport_manager().screen().screen_ndarray())
+            self.observation_space = Box(low=0, high=255, shape=screen.shape, dtype=np.uint32)
+        elif observation_type == 'tiles':
+            nvec = 384 * np.ones(self.game_wrapper.shape)
```
Comment (Baekalfen) [inline pyboy/gymboy.py:107]: > 384

Could be `pyboy.botsupport.constants.TILES` instead of a "magic number"

File: pyboy/pyboy.py [hunk 6]
```diff
@@ -221,6 +222,34 @@ def botsupport_manager(self):
             The manager, which gives easier access to the emulated game through the classes in `pyboy.botsupport`.
         """
         return botsupport.BotSupportManager(self, self.mb)
+    
+    def openai_gymboy(self, observation_type='tiles', buttons_press_mode='press', simultaneous_actions=False):
+        """
+        For Reinforcement learning, it is often easier to use the standard gym environment. This method will provide one.
+        The pyboy need a game_wrapper with a fitness function in order to build a reward system.
```
Comment (Baekalfen) [inline pyboy/pyboy.py:229]: > The pyboy need a game_wrapper with a fitness function in order to build a reward system.

Maybe change this to something like: "This function requires PyBoy to implement a Game Wrapper for the loaded ROM. You can find the supported games in `pyboy.plugins`."

File: pyboy/pyboy.py [hunk 7]
```diff
@@ -221,6 +222,34 @@ def botsupport_manager(self):
...
+                - 'tiles' gives the id of the sprites in 8x8 pixel zones of the game_area defined by the game_wrapper (Only useful in grid-based games).
+
+        buttons_press_mode: str
+            Define how the agent will interact with button inputs:
+                - 'press' the agent will only press inputs for 1 frame an then release it.
+                - 'toggle' the agent will toggle inputs, first time it press and second time it release.
+                - 'all' the agent have acces to all inputs, press and release are separated.
+
+        simultaneous_actions: bool
+            If true, the agent is allowed to inject multiple inputs at the same time. Caution, this also means that the action_space is way bigger (n -> 2^n)!
```
Comment (Baekalfen) [inline pyboy/pyboy.py:245]: It seems this documentation is present at three places. Maybe choose one place to have all the details, and refer to it from the other to places. So we don't have to keep all of them in sync.

File: pyboy/gymboy.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
+#
+# License: See LICENSE.md file
+# GitHub: https://github.com/Baekalfen/PyBoy
+#
+# File Contributors : MathisFederico 2020,
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py:5]: Okay !

File: pyboy/gymboy.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+import numpy as np
+
+from gym import Env
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py:9]: Yes it is ! It's called 'gym'

File: pyboy/gymboy.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+from gym.spaces import Discrete, MultiDiscrete, Box
+
+class GymBoy(Env):
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py:12]: Yeah sure ahah, it was a great name, but it more clear like that !

File: pyboy/gymboy.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    Arguments
+    ---------
+
+        game: `pyboy.PyBoy`
+            A PyBoy game instance
+
+        observation_type: str
+            Define what the agent will be able to see :
+                - 'raw' gives the raw pixels color
+                - 'tiles' gives the id of the sprites in 8x8 pixel zones of the game_area defined by the game_wrapper (Only useful in grid-based games).
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py:25]: I'm ok with both modifications, do I have to do them or will you ?

File: pyboy/gymboy.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+        buttons_press_mode: str
+            Define how the agent will interact with button inputs:
+                - 'press' the agent will only press inputs for 1 frame an then release it.
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py:29]: typos ... always a pleasure xD

File: pyboy/gymboy.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py:12]: In fact PyBoyGymEnv might be more precise

File: pyboy/gymboy.py (hunk 3 above)
Comment (MathisFederico) [reply pyboy/gymboy.py:93]: cleaner, i like it

File: pyboy/gymboy.py (hunk 4 above)
Comment (MathisFederico) [reply pyboy/gymboy.py:98]: Done

File: pyboy/gymboy.py (hunk 5 above)
Comment (MathisFederico) [reply pyboy/gymboy.py:107]: Sure ! Did not know that this constant was there !

File: pyboy/pyboy.py (hunk 6 above)
Comment (MathisFederico) [reply pyboy/pyboy.py:229]: Done

File: pyboy/pyboy.py (hunk 7 above)
Comment (MathisFederico) [reply pyboy/pyboy.py:245]: Not sure how to do that though :/

**Comment by MathisFederico** [comment]
> 
> 
> Really solid work! I'm impressed!
> 
> Don't be scared by all the comments. I do think you did a fine job!

Thank you ! If you realy think so, I would appreciate a recommendation on [my LinkedIn](https://www.linkedin.com/in/mathis-federico/) as it could help me in the futur !

**Comment by Baekalfen** [comment]
Sure, I can make a recommendation. Do you have some code you used to test the gym with? I  would like to have as much as possible in the repo covered by tests, so we know if something breaks.

Have you tried to train an AI with your code?

**Comment by MathisFederico** [comment]
> Have you tried to train an AI with your code?

Not yet, I need to finish my deepRL feature in LearnRL to have a proper test (because for now only table agent is stable but there is too much states (10,18,2) is 10^54 states ...)

But I'm working on it !

> Do you have some code you used to test the gym with?

It's pretty simple, just need to run a game with steps and make sure that : steps properly make a frame step, taking into account the action and action_type, done is True only when the game is done, rewards comes as expected,

For tests it would be nice to have a saved state at some points (for example clearing a row in tetris) to test rewards.

Shall I open an issue for that ?

**Comment by MathisFederico** [comment]
Also, it is hard for me to write tests because i didn't manage to get the compiler running, so when using new features the emulation is running realy slow ...

**Comment by Baekalfen** [comment]
Tell me when the PR is ready for review again.

**Comment by MathisFederico** [comment]
> Tell me when the PR is ready for review again.

Now I think it is !
I just don't know how to test the rewards correctly and fast ...
(For my tests I cleared the first row of tetris and checked that I won a 40 reward at that frame but it would be slow to do that ...)

File: pyboy/gymboy.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                - 'toggle' the agent will toggle inputs, first time it press and second time it release.
+                - 'all' the agent have acces to all inputs, press and release are separated.
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py]: It was more "will have access" I think

File: pyboy/gymboy.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+        
+        # Build game_wrapper
+        self.game_wrapper = game.game_wrapper()
+        if self.game_wrapper is None:
+            raise ValueError("You need to build a game_wrapper to use this function. Otherwise there is no way to build a reward function automaticaly (for now).")
+        self.last_fitness = self.game_wrapper.fitness
+
+        # Building the action_space
+        self._DO_NOTHING = -1
+        self._buttons = list(range(1, 9)) # UP DOWN RIGHT LEFT A B SELECT START
```
[the remark this answers is not in the document]
Comment (MathisFederico) [reply pyboy/gymboy.py]: Yeah, I've done that before I had the pyboy circular import where i removed all imports, but WindowEvent is fine !

It's changed back to that now !

**Comment by MathisFederico** [comment]
> Tell me when the PR is ready for review again.

Up for PR review now !

I miss some tests on reward before cleaning the first row is hard to hard-code from the begining and slow for a test ...

**Comment by Baekalfen** [comment]
I'll wait with the review until the tests are done

File: tests/test_openai_gym.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+import numpy as np
+import pytest
+from pyboy import PyBoy, WindowEvent
+from pyboy.botsupport.constants import COLS, ROWS
+
+try:
+    from .utils import tetris_rom
+except:
+    tetris_rom = "ROMs/Tetris.GB"
```
Comment (Baekalfen) [inline tests/test_openai_gym.py:14]: The wrong Tetris ROM could have been caught here. `utils.tetris_rom` is checksum based, and doesn't recognize your ROM as the correct Tetris ROM

**Comment by MathisFederico** [comment]
Finished the test for everything except done because it's a one liner from game_wrapper.game_over() thus Closing #129

**Comment by Baekalfen** [comment]
Awesome work! I've very satisfied with the result!

**Comment by MathisFederico** [comment]
> Awesome work! I've very satisfied with the result!

Thank you very much ! It was really nice working with you ! I'm looking for internships starting in september 2020 so if you really think that I did a good job, a recommendation on [LinkedIn](https://www.linkedin.com/in/mathis-federico/) would be much appreciated !

This weekend I will try to make a stable version on [LearnRL](https://github.com/MathisFederico/LearnRL) so that anyone can build his own IA on PyBoy with only few knowledge about RL and DeepLearning !

**Pull request merged**